Kubeadm - Add timeouts to kubeadm config API

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR adds to the kubeadm config API the possibilty to configure timeouts.

As discussed at sig level, the kubeadm config API will be designed for supporting configuration of many timeouts, but only the "waiting for the control-plane to start" timetout will be implemented ATM.

**Which issue(s) this PR fixes** : 
Fixes: https://github.com/kubernetes/kubeadm/issues/1168

**Special notes for your reviewer**:
This is a WIP, aimed at collecting feedbacks on the API design.

/sig cluster-lifecycle
/kind documentation
/assign @timothysc
/assign @dtieber
/cc @rosti  
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@neolit123 
@chuckha 

**Release note**:
```release-note
Kubeadm config API now allows configuration of the timeout for waiting the control-plane to start both in the kubeadm init and kubeadm join --xontrol-plane workflow
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:
```diff
@@ -47,6 +47,9 @@ type InitConfiguration struct {
 	APIEndpoint APIEndpoint `json:"apiEndpoint,omitempty"`
+
+	// Timeouts defines maximum time intervals kubeadm should wait for completion of critical steps of the init workflow.
+	Timeouts map[KubeadmTimeout]metav1.Duration
```
i think this way of mapping the timeouts will work fine and it would be easy to use from YAML.
+1

timothysc replied:
map? What other timeouts do we want to expose to the users?  

/cc @mauilion

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:
```diff
@@ -354,3 +360,13 @@ type AuditPolicyConfiguration struct {
+	// WaitForControlPlaneTimeout indicates the timeout duration kubeadm should wait when waiting for the control-plane
+	// to start in the kubeadm init workflow and in the kubeadm join controlplane workflow
+	// defaults to 4 minutes
+	WaitForControlPlaneTimeout = "controlplane"
```
a question here would be, what other types of timeouts we want to add for the init and join configurations?

rosti replied:
@neolit123 possibly discovery timeout, although I highly recommend keeping the original spot it occupies (`JoinConfiguration.Discovery.Timeout`)

neolit123 reviewed (commented):
+1 thanks @fabriziopandini

rosti commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:
```diff
@@ -274,6 +277,9 @@ type JoinConfiguration struct {
 	// FeatureGates enabled by the user.
 	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+
+	// Timeouts defines maximum time intervals kubeadm should wait for completion of critical steps of the init workflow.
```
s/init workflow/join workflow

rosti commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:
```diff
@@ -274,6 +277,9 @@ type JoinConfiguration struct {
 	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+
+	// Timeouts defines maximum time intervals kubeadm should wait for completion of critical steps of the init workflow.
+	Timeouts map[KubeadmTimeout]metav1.Duration
```
We currently have a discovery timeout `Discovery.Timeout`, should we move or link that in here as another key?

rosti reviewed (commented):
Looks OK to me.
I wonder if it would be more handy to implement timeouts on a per-phase basis? The keys of the `Timeouts` field would be phase names and the phase runner will be responsible for monitoring the timeouts.

timothysc reviewed (commented):
/approve

timothysc commented:
@fabriziopandini we can chat about this tomorrow, but I'm fine with a single timeout, just for the apiserver.

fabriziopandini committed:
add timeouts to kubeadm config

fabriziopandini committed:
autogenerated

fabriziopandini commented:
@timothysc @neolit123 @rosti 
Removed WIP and ready for another pass

> @timothysc  What other timeouts do we want to expose to the users?
 
DIscovery timeouts 

> @rosty:  although I highly recommend keeping the original spot it occupies (JoinConfiguration.Discovery.Timeout)

I think that we should have a unique method of settings all timeouts, so I removed from the original spot. Of course open to review it if there is no consensus

timothysc commented:
In discussion today, we will roll in a control plane timeout into the other component work for v1beta1 

/hold

fabriziopandini commented:
I'm in contact with @rosti for addressing the control-plane timeout in another PR
/close

migueleliasweb commented:
Hey guys, sorry to post in such an old issuebut I have the exact same problem. The Api server takes a little bit too long for kubeadm to say it's ready. With the timeout config, this would get solved easily.

I couldn't find the other PR that would superseed this one. Was it merged? Can someone post a link to it?

Cheers!

neolit123 commented:
there is a control plane timeout in the [v1beta1 config](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1):
```
# config.yaml
apiVersion: kubeadm.k8s.io/v1beta1
kind: ClusterConfiguration
apiServer:
  timeoutForControlPlane: 4m0s
```

```
kubeadm init ... --config config.yaml
```
